src/common/extensions/array.ts

```diff
@@ -0,0 +1,61 @@
+import Guard from "../guard";
+
+/**
+ * Processes items in the array within the specified batch size (default: 5)
+ * @param this The array to process
+ * @param action The action to perform on each item in the array
+ * @param batchSize The batch size for actions to perform in parallel (default: 5)
+ */
+export async function forEachAsync<T>(
+    this: T[],
+    action: (item: T) => Promise<void>,
+    batchSize: number = 5): Promise<void> {
+    Guard.null(this);
+    Guard.null(action);
+    Guard.expression(batchSize, (value) => value > 0);
+
+    const all: T[] = [...this];
```

Comment:
As this is a generic extension for any array and potentially it could be reused in other cases with different requirements (i.e. array and/or element size) I just wonder why we copy the original array on a mutable variable instead of directly using the "this" array with a index instead of pop() from this copy.

Reply:
`pop()` from the original array would mutate the original array which would be an unintended side effect.

Reply:
@wbreza of course.  Indeed I was suggesting to use an index iterator and do not mutate neither the original or the copy